src/main/scala/com/fulcrumgenomics/bam/IdentifyPrimers.scala

```diff
@@ -0,0 +1,1027 @@
+            alignmentResult.score >= minAlignmentScore && alignmentResult.queryEnd >= minQueryEnd
+          }
+          // now just keep the primer and score
+          .map { case (alignmentInput, alignmentResult) => PrimerAndScore(alignmentInput.query, alignmentResult.score) }
+          // get the best two alignment results by score
+          .maxTwoBy(_.score)
+
+        // Create a primer match if alignments were found (i.e. a Option[PrimerMatch])
+        alignmentResults match {
+          case (None, _)                    => None
+          case (Some(best), None)           => Some(FullAlignmentPrimerMatch(primer = best.primer, score = best.score, secondBestScore = minAlignmentScore))
+          case (Some(best), Some(nextBest)) => Some(FullAlignmentPrimerMatch(primer = best.primer, score = best.score, secondBestScore = nextBest.score))
+        }
+    }
+  }
+
+  /** Processes a single batch of templates. */
+  def processBatch(templates: Seq[Template],
+                   metricCounter: SimpleCounter[TemplateTypeMetric],
+                   progress: ProgressLogger): Seq[Template] = {
+
+    val counter = new SimpleCounter[TemplateTypeMetric]()
+    val aligner  = newAligner
+
+    templates.foreach { template =>
+      // match up to but not including full alignment
+      val r1MatchOption = template.r1.map { r1 => fwdMatcher.matchUntilAlignment(r1) }
+      val r2MatchOption = template.r2.map { r1 => revMatcher.matchUntilAlignment(r1) }
+
+      // full in the alignment
+      val r1Match = completePrimerMatch(r1MatchOption, aligner)
+      val r2Match = completePrimerMatch(r2MatchOption, aligner)
+
+      // add the three prime matching
+      val r1ThreePrimeMatch = template.r1.flatMap { r1 => matchThreePrime(r1, r1Match, fwdMatch=true) }
+      val r2ThreePrimeMatch = template.r2.flatMap { r2 => matchThreePrime(r2, r2Match, fwdMatch=false) }
+
+      // Get information about the matches
+      val templateTypes = (template.r1, template.r2) match {
+        case (Some(r1), Some(r2)) =>
+          val rType = (r1.mapped, r2.mapped) match {
+            case (true, true)   => TemplateType.MappedPair
+            case (true, false)  => TemplateType.Unpaired
+            case (false, true)  => TemplateType.Unpaired
+            case (false, false) => TemplateType.UnmappedPair
+          }
+          val mType = formatPair(r1, r2, r1Match, r2Match, r1ThreePrimeMatch, r2ThreePrimeMatch)
+          TemplateTypeMetric(rType, mType, r1Match, r2Match)
+        case (Some(r1), None) =>
+          require(!r1.paired, s"Found paired read but missing R2 for ${r1.name}")
+          val rType = if (r1.mapped) TemplateType.MappedFragment else TemplateType.UnmappedFragment
+          val mType = formatFragment(r1, r1Match, r1ThreePrimeMatch)
+          TemplateTypeMetric(rType, mType, r1Match, r2Match)
+        case _ =>
+          unreachable(s"Template did not have an R1: ${template.name}")
```

Comment:
I don't think I'd use `unreachable` here.  Probably just `throw new IllegalStateException`, as users are plenty capable of creating BAMs that are missing R1s, and `unreachable` is meant for places that are truly unreachable.